### DISCUSSION TITLE
Matching first line of large code-block

### DISCUSSION THREAD

**Discussion by Abhiram98** [ast-grep/ast-grep#705] [Q&A] [answered]
Firstly, thanks for the awesome tool!

I've been trying to use the tool to match python code fragments(~5-10 loc), while allowing multiple random statements in between. I've been trying to use the "preceeds" relational rule, along with a "stopBy:end" condition, but that doesn't seem to be working.

Here is a [link](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InB5dGhvbiIsInF1ZXJ5IjoiY29uc29sZS5sb2coJE1BVENIKSAgICAgIDtcblxuXG4iLCJyZXdyaXRlIjoiIiwiY29uZmlnIjoiIyBZQU1MIFJ1bGUgaXMgbW9yZSBwb3dlcmZ1bCFcbiMgaHR0cHM6Ly9hc3QtZ3JlcC5naXRodWIuaW8vZ3VpZGUvcnVsZS1jb25maWcuaHRtbCNydWxlXG5ydWxlOlxuICBwYXR0ZXJuOiBcbiAgICBjb250ZXh0OiAkQ09VTlQ9MFxuICAgIHNlbGVjdG9yOiBleHByZXNzaW9uX3N0YXRlbWVudFxuICBwcmVjZWRlczpcbiAgICBwYXR0ZXJuOiBcbiAgICAgIGNvbnRleHQ6IFwiZm9yICRJIGluICRBUlI6ICRDT1VOVCs9JElcIlxuICAgICAgc2VsZWN0b3I6IGZvcl9zdGF0ZW1lbnRcbiAgICBzdG9weUJ5OiBlbmRcblxuIiwic291cmNlIjoiY291bnQgPSAwXG5zPTErMlxueD0xXG5mb3IgaSBpbiBhcnJheTpcbiAgICBjb3VudCArPSBpXG4ifQ==) to the playground I've been working with.

Here is the code-block I would like to search for 
```
count = 0
for i in array:
    count += i
```

Here are some code-blocks that I would like to match:

```
count = 0
<random line(s) of code>
for i in array:
    <random line(s) of code>
    count += i
```

```
count = 0
s=1+2
x=1
for i in array:
    y = 124
    count += i
```

**Comment by HerringtonDarkholme** [comment]
Sorry I didn't quite get what your error message means. I guess you want to [match a variable that is incremented in a for loop](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InB5dGhvbiIsInF1ZXJ5IjoiJENPVU5UID0gJFxuIiwicmV3cml0ZSI6IiIsImNvbmZpZyI6InJ1bGU6XG4gIHBhdHRlcm46IFxuICAgIGNvbnRleHQ6ICRDT1VOVCA9ICRfXG4gICAgc2VsZWN0b3I6IGV4cHJlc3Npb25fc3RhdGVtZW50XG4gIHByZWNlZGVzOlxuICAgIGtpbmQ6IGZvcl9zdGF0ZW1lbnRcbiAgICBoYXM6XG4gICAgICBwYXR0ZXJuOiAkQ09VTlQgKz0gJF9cbiAgICAgIHN0b3BCeTogZW5kXG4gICAgc3RvcEJ5OiBlbmRcbiIsInNvdXJjZSI6ImNvdW50ID0gMFxucz0xKzJcbng9MVxuZm9yIGkgaW4gYXJyYXk6XG4gICAgY291bnQgKz0gaVxuIn0=)?

It it was the case, this [playground link](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InB5dGhvbiIsInF1ZXJ5IjoiJENPVU5UID0gJFxuIiwicmV3cml0ZSI6IiIsImNvbmZpZyI6InJ1bGU6XG4gIHBhdHRlcm46IFxuICAgIGNvbnRleHQ6ICRDT1VOVCA9ICRfXG4gICAgc2VsZWN0b3I6IGV4cHJlc3Npb25fc3RhdGVtZW50XG4gIHByZWNlZGVzOlxuICAgIGtpbmQ6IGZvcl9zdGF0ZW1lbnRcbiAgICBoYXM6XG4gICAgICBwYXR0ZXJuOiAkQ09VTlQgKz0gJF9cbiAgICAgIHN0b3BCeTogZW5kXG4gICAgc3RvcEJ5OiBlbmRcbiIsInNvdXJjZSI6ImNvdW50ID0gMFxucz0xKzJcbng9MVxuZm9yIGkgaW4gYXJyYXk6XG4gICAgY291bnQgKz0gaVxuIn0=) should work.

```yml
rule:
  pattern: 
    context: $COUNT = $_
    selector: expression_statement
  precedes:
    kind: for_statement
    has:
      pattern: $COUNT += $_
      stopBy: end
    stopBy: end

```

**Reply by Abhiram98** [reply]
I would like to search for this entire code block:

```
count = 0
for i in array:
    count += i
```

And I would like the search to find matches even if there are some additional statements in between the `count=0` and the for loop.

**Reply by HerringtonDarkholme** [reply]
Nope, this is currently impossible to match multiple nodes using YAML now. You can try napi/python API instead.

**Reply by Abhiram98** [reply]
I don't mind matching just the first statement `count=0`. [This](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InB5dGhvbiIsInF1ZXJ5IjoiJENPVU5UID0gJFxuIiwicmV3cml0ZSI6IiIsImNvbmZpZyI6InJ1bGU6XG4gIHBhdHRlcm46IFxuICAgIGNvbnRleHQ6ICRDT1VOVCA9IDBcbiAgICBzZWxlY3RvcjogZXhwcmVzc2lvbl9zdGF0ZW1lbnRcbiAgcHJlY2VkZXM6XG4gICAgcGF0dGVybjogXG4gICAgICBjb250ZXh0OiA+XG4gICAgICAgIGZvciAkSSBpbiAkQVJSOiBcblxuICAgICAgICAgICQkJFNcbiAgICAgICAgICBcbiAgICAgICAgICAkQ09VTlQrPSRJXG4gICAgICBzZWxlY3RvcjogZm9yX3N0YXRlbWVudFxuICAgIHN0b3BCeTogZW5kXG4iLCJzb3VyY2UiOiJjb3VudCA9IDBcbng9MVxuZm9yIGkgaW4gYXJyYXk6XG4gICAgIyB0aGlzIGNvbW1lbnQgYnJlYWtzIHRoZSBzZWFyY2hcbiAgICBjb3VudCArPSBpXG4ifQ==) is kind of what I'm going for, but a comment seems to break the search?

**Reply by HerringtonDarkholme** [reply]
Python tree-sitter gives different output so the pattern needs to change.
<img width="433" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/6fd3c37c-dc8d-4bf4-931b-bc031ac42bf5">

You can also inspect the AST in the playground.  [playground manual](https://ast-grep.github.io/reference/playground) is on the top of page. 
<img width="507" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/f73bf4df-54f8-4ee0-b145-6f066b92afb5">

**Comment by HerringtonDarkholme** [comment]
If you really need to report multiple lines without using programmatic API. You can use JSON mode.

```
id: test
language: python
rule:
  pattern:
    context: $COUNT = $_
    selector: expression_statement
  precedes:
    pattern: $LOOP
    kind: for_statement
    has:
      pattern: $COUNT += $_
      stopBy: end
    stopBy: end
```

```shell
sg scan -r test.yml test.py --json | jq '.[].metaVariables.single'
```
Which gives you  the start line of `COUNT` and the end line of `LOOP`
<img width="669" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/c84ed190-5cf8-4f0d-a400-3d134b3f4358">

**Comment by Abhiram98** [comment, marked as answer]
Thanks for your replies, they were helpful! [This](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InB5dGhvbiIsInF1ZXJ5IjoiJENPVU5UID0gJFxuIiwicmV3cml0ZSI6IiIsImNvbmZpZyI6InJ1bGU6XG4gIHBhdHRlcm46XG4gICAgY29udGV4dDogJENPVU5UPTBcbiAgICBzZWxlY3RvcjogZXhwcmVzc2lvbl9zdGF0ZW1lbnRcbiAgcHJlY2VkZXM6XG4gICAga2luZDogZm9yX3N0YXRlbWVudFxuICAgIGFsbDpcbiAgICAgIC0gaGFzOlxuICAgICAgICAgIGZpZWxkOiBsZWZ0XG4gICAgICAgICAgcGF0dGVybjogJElcbiAgICAgIC0gaGFzOlxuICAgICAgICAgIGZpZWxkOiByaWdodFxuICAgICAgICAgIHBhdHRlcm46ICRBUlJcbiAgICAgIC0gaGFzOlxuICAgICAgICAgIHBhdHRlcm46ICRDT1VOVCs9JElcbiAgICAgICAgICBzdG9wQnk6IGVuZFxuICAgIHN0b3BCeTogZW5kIiwic291cmNlIjoiY291bnQgPSAwXG54PTFcbmZvciBpIGluIGFycmF5OlxuICAgIGNvdW50ICs9IGkifQ==) is what I ended up with

**Reply by HerringtonDarkholme** [reply]
Nice! Thanks for trying out ast-grep!
